libV2/client/index.js

```diff
@@ -84,6 +89,11 @@ class UtapiClient {
         this._drainCanSchedule = true;
         this._drainDelay = (config && config.drainDelay) || 30000;
         this._suppressedEventFields = (config && config.suppressedEventFields) || null;
+        const eventFilters = (config && config.filter) || {};
+        this._shouldPushMetric = buildFilterChain(eventFilters);
```

Comment:
I'm not sure if building a filter chain like done isn't overengineered, instead of a simpler existence check of keys in two sets, allowed and denied, pre-filled from the configuration (or just `denied` since `allowed` may not be needed per the requirements)

Something like

```
for each key passed to the filter function
    if there is an `allow` set
        if the key exists in the set, return "allow"
        else return "deny"
    else if there is a "deny" set
        if the key exists in the set, return "deny"
        else return "allow"

Reply:
While I agree that the chain is more complex that a loop, I think that avoiding the overhead of the loop and checks at runtime is worth it given that this will be called on most API calls.

Reply:
Correct me if I'm wrong, what I understand that we are mostly saving is the `if` condition on whether we should check on the allow or deny set since it is memoized by creating an ad-hoc function on either of the sets. We are also indeed not using a loop over the keys, but replacing it by nested calls to the ad-hoc function, which I'm not convinced is more efficient than a loop since function calls also have a cost in themselves.

Overall I feel we are trying to save on crumbs at the expense of complexity and we may not even achieve the intended savings, hence I believe it is best not to try to optimize at this point, since `Set` structures should be efficient enough for the need and I believe should not incur a measurable overhead (but maybe I'm wrong, just my feeling).

Reply:
We are actually not saving on the lookups in the `Set`, but my point is that with a simple loop we should get the expected performance anyways, we can always revisit this if we observe a noticeable impact.

One random idea if that turns out to be the case that could be a more efficient optimization is using a bloom filter on the set of keys that we want to deny, so that we can check quickly if none of them is in the set with high probability.